src/routes/docs.js

```diff
@@ -50,6 +61,42 @@ router.get('/:page', async (request, response, next) => {
 	response.render('docs/docs', renderData);
 });
 
+router.get([
+	'/errors/:errorCode', // TODO - Deprecate "errors" route
+	'/error/:errorCode',
+], async (request, response, next) => {
+	const locale = response.locals.localeString;
+	const error = request.params.errorCode;
+
+	const [ sysmodule, errorCode ] = error.split('-');
+
+	if (!errorCode) {
+		return next();
+	}
+
+	let template = fs.readFileSync(path.join(__dirname, '../../docs/common/error-page-template.md'), {
+		encoding: 'utf8'
+	});
+
+	const errorInfo = errors.getErrorInfo(sysmodule, errorCode, locale);
+
+	if (!errorInfo) {
+		return next();
+	}
+
+	template = template.replace('{module}', sysmodule);
```

Comment:
This will iterate the whole template string for each replace and will only replace a single instance of each of the placeholders

I'd recommend taking a look at the [express markdown example](https://github.com/expressjs/express/blob/master/examples/markdown/index.js#L17-L25) where they perform a single replace and use a function to look up the right data. Something like this could be added to the `util.parseDoc` maybe?

Reply:
Correct, it will iterate the whole string. However it's an very small string, and the operations are extremely fast

Because of this I prioritized readability over trying to micro-optimize these operations

However to give some numbers:

On a test of 200 iterations the string-replace method took on average 0.016665ms to complete, between 0.008ms and 0.143ms. Using the same kind of regex-replace method as shown in the example you linked (which was what I initially did use), is actually slower in this case. In the same sample of 200 iterations it took an average of 0.02626ms to complete, between 0.013ms and 0.265ms

Reply:
> Correct, it will iterate the whole string. However it's an very small string, and the operations are extremely fast
Totally agree, was more pointing it out rather than saying it wouldn't be performant

I think my main point was that it'll only replace a single instance of each of the placeholders. This totally works for the current template, but if you ever added the same placeholder in another spot in the markdown that would require you update the code here too. I appreciate that might be incredibly unlikey and a good case for YAGNI, but thought I'd mention it 👍

Reply:
Of course, even if we ultimately disagree with a suggestion please do feel free to still make them :+1: I value all feedback and suggestions, even ones we end up not going with

Breaking this out into a more generic system is not a *bad* idea, and in fact may be something to consider again if we were to add these kinds of translation templates for other pages (such as the docs pages). But currently yes this is a good case for YAGNI, we only need this kind of template system for one page and this was the simplest/fastest way to get that done

Though I did misunderstand your suggestion originally. I assumed you made it for performance reasons, due to your previous comment and because you brought up to amount of passes through the string, so that was one me

Also it should be noted that the current state of the website is, for all intents and purposes, temporary. @hauntii is working on a total rewrite of the website, migrating it to TypeScript and using Next rather than express+handlebars. So all of this will be trashed eventually anyway